# Belief in Supernatural Evil and Guns

The purpose of this assignment is to attempt to replicate the follwing paper, which examines the relationship of belief in a supernatural evil metric (as well as of other variables) with some gun policies.

---

> Natalia Katsiapi, 810040 <br/>
> Department of Management Science and Technology <br/>
> Athens University of Economics and Business <br/>
> t8180040@aueb.gr

In [11]:
import pandas as pd
import sklearn as sk
import numpy as np
import statsmodels.formula.api as smf
from statsmodels.formula.api import logit

from plotnine import *
%matplotlib inline
pd.set_option('display.max_columns', None)

The dataset is **Baylor Religion Survey, Wave IV (2014)** and was downloaded from [here](https://www.thearda.com/Archive/Files/Descriptions/BRS2014.asp).

In [12]:
df = pd.read_excel("Baylor Religion Survey, Wave IV (2014).XLSX", sheet_name=0)
print(df.shape)
df.head()

(1572, 282)


MOTHERLODE_ID  PROJECT_ID       RESPONDENT_ID     METHOD_ID METHOD_TYPE  \
0      128170892       33370  01_000001_00000001  162048140_01        SCAN   
1      128170893       33370  01_000002_00000001  162048140_01        SCAN   
2      128170894       33370  01_000003_00000001  162048140_01        SCAN   
3      128170895       33370  01_000004_00000001  162048140_01        SCAN   
4      128170896       33370  01_000005_00000001  162048140_01        SCAN   

   PROJECT_NUMBER CREATED_ON PRACTICE RESPONDENT_DATE RESPONDENT_LANGUAGE  \
0       162048140  3/26/2014    OTHER       1/24/2014               en-US   
1       162048140  3/26/2014    OTHER       1/24/2014               en-US   
2       162048140  3/26/2014    OTHER       1/24/2014               en-US   
3       162048140  3/26/2014    OTHER       1/24/2014               en-US   
4       162048140  3/26/2014    OTHER       1/24/2014               en-US   

   TEMPLATE_NUMBER_162048140   ENTITY_ID  SCAN_RESPONDENT_ID  AUTO_INCREMENT  \
0                  162048140  4139835441          4139835441               1   
1                  162048140  4139832983          4139832983               2   
2                  162048140  4139835115          4139835115               3   
3                  162048140  4139837428          4139837428               4   
4                  162048140  4139835007          4139835007               5   

   CASEID  BATCH_ID  LANG1    Q1  Q2_DK   Q3  Q3_1   Q4   Q7   Q8  Q25  Q10  \
0       1         1  en-US  20.0    NaN  4.0   4.0  8.0  5.0  1.0  3.0  5.0   
1       2         1  en-US   NaN    NaN  4.0   4.0  1.0  1.0  8.0  3.0  3.0   
2       3         1  en-US  31.0    NaN  3.0   4.0  7.0  5.0  5.0  3.0  5.0   
3       4         1  en-US  33.0    NaN  3.0   4.0  6.0  4.0  2.0  3.0  4.0   
4       5         1  en-US  14.0    NaN  4.0   4.0  8.0  5.0  2.0  2.0  5.0   

   Q11  Q13A  Q13B  Q13C  Q13D  Q13E  Q13F  Q13G  Q13H  Q13I  Q13J  Q13K  \
0  8.0   2.0   4.0   1.0   1.0   2.0   1.0   1.0   1.0   1.0   1.0   1.0   
1  5.0   1.0   1.0   1.0   2.0   2.0   2.0   1.0   1.0   1.0   1.0   1.0   
2  1.0   2.0   1.0   1.0   1.0   3.0   2.0   1.0   1.0   1.0   1.0   1.0   
3  6.0   3.0   2.0   2.0   2.0   2.0   2.0   2.0   2.0   1.0   1.0   1.0   
4  8.0   4.0   4.0   1.0   2.0   4.0   4.0   1.0   1.0   1.0   1.0   1.0   

   Q13L  Q14A  Q14B  Q14C  Q14D  Q14E  Q14F  Q14G  Q17  Q18   Q9  Q19A  Q19B  \
0   1.0   2.0   2.0   1.0   1.0   NaN   1.0   3.0  1.0  1.0  5.0   4.0   4.0   
1   1.0   8.0   8.0   8.0   8.0   8.0   8.0   8.0  1.0  1.0  1.0   4.0   3.0   
2   1.0   1.0   3.0   1.0   0.0   1.0   0.0   3.0  2.0  1.0  3.0   3.0   2.0   
3   1.0   3.0   2.0   1.0   1.0   1.0   2.0   3.0  2.0  1.0  2.0   4.0   4.0   
4   1.0   4.0   0.0   0.0   1.0   0.0   1.0   1.0  1.0  1.0  4.0   4.0   4.0   

   Q19C  Q19D  Q19E  Q19F  Q20A  Q20B  Q20C  Q20D  Q20E  Q20F  Q20G  Q20H  \
0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   
1   4.0   4.0   4.0   3.0   4.0   1.0   4.0   3.0   4.0   NaN   4.0   NaN   
2   3.0   3.0   3.0   2.0   4.0   2.0   4.0   1.0   4.0   1.0   4.0   1.0   
3   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   
4   4.0   4.0   4.0   4.0   4.0   2.0   4.0   1.0   4.0   2.0   4.0   2.0   

   Q21  Q23A  Q23B  Q23C  Q23D  Q23E  Q23F  Q23G  Q24A  Q24B  Q24C  Q24D  \
0  2.0   4.0   4.0   4.0   1.0   4.0   4.0   4.0   4.0   1.0   4.0   4.0   
1  2.0   4.0   4.0   4.0   1.0   4.0   4.0   4.0   NaN   1.0   2.0   NaN   
2  4.0   3.0   3.0   2.0   2.0   2.0   3.0   2.0   3.0   2.0   3.0   2.0   
3  4.0   4.0   4.0   4.0   2.0   3.0   4.0   4.0   4.0   2.0   4.0   4.0   
4  2.0   4.0   4.0   4.0   4.0   4.0   4.0   4.0   NaN   1.0   NaN   4.0   

   Q24E  Q24F  Q24G  Q26  Q27A  Q27B  Q27C  Q27D  Q27E  Q28A  Q28B  Q28C  Q29  \
0   4.0   1.0   4.0  4.0   1.0   1.0   1.0   1.0   1.0   4.0   4.0   4.0  1.0   
1   4.0   NaN   4.0  4.0   1.0   1.0   2.0   1.0   1.0   2.0   3.0   2.0  2.0   
2   3.0   2.0 

We will gain a general perspective of the data.

In [13]:
df.describe()

MOTHERLODE_ID    PROJECT_ID  PROJECT_NUMBER  TEMPLATE_NUMBER_162048140  \
count   1.572000e+03   1572.000000          1572.0                     1572.0   
mean    1.281718e+08  33370.021628     162048140.0                162048140.0   
std     5.894006e+02      0.145513             0.0                        0.0   
min     1.281709e+08  33370.000000     162048140.0                162048140.0   
25%     1.281713e+08  33370.000000     162048140.0                162048140.0   
50%     1.281717e+08  33370.000000     162048140.0                162048140.0   
75%     1.281722e+08  33370.000000     162048140.0                162048140.0   
max     1.281729e+08  33371.000000     162048140.0                162048140.0   

          ENTITY_ID  SCAN_RESPONDENT_ID  AUTO_INCREMENT       CASEID  \
count  1.572000e+03        1.572000e+03     1572.000000  1572.000000   
mean   4.139833e+09        4.139833e+09      831.400763   831.400763   
std    4.662307e+03        4.662307e+03      522.917758   522.917758   
min    4.139818e+09        4.139818e+09        1.000000     1.000000   
25%    4.139832e+09        4.139832e+09      376.750000   376.750000   
50%    4.139834e+09        4.139834e+09      792.500000   792.500000   
75%    4.139836e+09        4.139836e+09     1255.250000  1255.250000   
max    4.139838e+09        4.139838e+09     1864.000000  1864.000000   

          BATCH_ID           Q1  Q2_DK           Q3         Q3_1           Q4  \
count  1572.000000  1550.000000  254.0  1553.000000  1553.000000  1556.000000   
mean     33.784351    91.458710    1.0     2.816484     3.337411     3.676735   
std      20.930729   798.668391    0.0     1.221241     1.089783     2.932510   
min       1.000000     1.000000    1.0     1.000000     1.000000     0.000000   
25%      16.000000    16.000000    1.0     2.000000     3.000000     1.000000   
50%      32.000000    25.000000    1.0     3.000000     3.000000     3.000000   
75%      51.000000    45.000000    1.0     4.000000     4.000000     7.000000   
max      75.000000  9998.000000    1.0     8.000000     8.000000     8.000000   

                Q7           Q8          Q25          Q10          Q11  \
count  1143.000000  1147.000000  1549.000000  1551.000000  1540.000000   
mean      3.311461     3.154316     3.099419     2.900064     3.197403   
std       1.475759     2.170661     1.487391     1.832240     3.139593   
min       1.000000     1.000000     1.000000     0.000000     0.000000   
25%       2.000000     2.000000     2.000000     1.000000     0.000000   
50%       3.000000     2.000000     3.000000     3.000000     2.000000   
75%       5.000000     5.000000     3.000000     5.000000     6.000000   
max       5.000000     8.000000     8.000000     5.000000     8.000000   

              Q13A         Q13B         Q13C         Q13D         Q13E  \
count  1496.000000  1480.000000  1462.000000  1458.000000  1487.000000   
mean      1.500668     1.558784     1.184679     1.235254     1.717552   
std       0.802171     0.951454     0.582741     0.614572     0.964153   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     2.000000     1.000000     1.000000     2.000000   
max       4.000000     4.000000     4.000000     4.000000     4.000000   

              Q13F         Q13G         Q13H         Q13I         Q13J  \
count  1460.000000  1460.000000  1462.000000  1462.000000  1465.000000   
mean      1.376712     1.136301     1.092339     1.073187     1.096928   
std       0.739935     0.484068     0.407438     0.346261     0.367977   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     1.000000     1.000000  

## Belief in Supernatural Evil Metric
Using the [codebook](https://www.thearda.com/Archive/Files/Codebooks/BRS2014_CB.asp) the questions needed for the metric are:

* Whether the respondent believes in the devil: Q23A

* Whether the respondent believes in hell: Q23C

* Whether the respondent believes in demons: Q23G
 

In [14]:
sup_evil = ['Q23A', 'Q23C', 'Q23G']
df[sup_evil].describe()

Q23A         Q23C         Q23G
count  1504.000000  1499.000000  1492.000000
mean      3.149601     3.088726     3.016086
std       1.108048     1.114655     1.108353
min       1.000000     1.000000     1.000000
25%       2.000000     2.000000     2.000000
50%       4.000000     4.000000     3.000000
75%       4.000000     4.000000     4.000000
max       4.000000     4.000000     4.000000

The metric the paper or is the mean of the answers to this question for each respondent. We will create a helper dataframe with the useful for the analysis variables.

In [15]:
selected_vars = pd.DataFrame()
selected_vars["supernatural_evil"] = df[sup_evil].mean(axis=1)
selected_vars["supernatural_evil"].describe()

count    1517.000000
mean        3.089541
std         1.053424
min         1.000000
25%         2.333333
50%         3.666667
75%         4.000000
max         4.000000
Name: supernatural_evil, dtype: float64

The overall mean indeed agrees with that of the paper.

In [16]:
df[sup_evil].mean(axis=1).mean()

3.08954076027247

To justify our choice to calculate the metric this way, we will use PCA, but first, in order for it to work, we will impute Nans.

In [17]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

imp = SimpleImputer(missing_values=np.nan, strategy='mean')
df[sup_evil] = imp.fit_transform(df[sup_evil])

scaler = StandardScaler()
X_s = scaler.fit_transform(df[sup_evil])
pca = PCA(n_components=1)
df['supernatural_evil'] = pca.fit_transform(X_s)
print(f'explained variance ratio: {pca.explained_variance_ratio_}')

explained variance ratio: [0.8976102]


The explained variance is high, thus projecting the answers of 3 questions to one metric is effective.

In [18]:
pca.components_

array([[-0.58620261, -0.57641524, -0.56930833]])

Thus, the supernatural evil metric is calculated for each respondent as follows:

$$y = -0.58620261 \times x_1 - 0.57641524 \times x_2 -0.56930833 \times x_3 $$


In the above: $x1$ is `Q23A` (devil), $x2$ is `Q23C` (hell) and $x3$ is `Q23G` (evil).
We can conclude that mean is a suitable measure, since the 3 weights are similar. Thus, there is no need to put different weights on the 3 variables, for example double results of `Q23A` and each variable equally contributes to the metric.

## Variables Selection


We will start by encoding the dependent variables' results to 0 and 1.

In [19]:
dependent_variables = ["semi_auto_guns", "ammo_clips", "handguns", "carry_laws", \
    "school_security", "teachers_guns", "gun_safety", "mental_health_screening"]
selected_vars[dependent_variables] = df[["Q73A", "Q73E", "Q73G", "Q73H", "Q73C",\
     "Q73F", "Q73B", "Q73D"]].apply(lambda x: x.map({1: 1, 2: 0}))
selected_vars[dependent_variables].describe()

semi_auto_guns   ammo_clips     handguns   carry_laws  school_security  \
count     1502.000000  1502.000000  1501.000000  1502.000000      1510.000000   
mean         0.629161     0.671105     0.290473     0.557923         0.621854   
std          0.483190     0.469968     0.454132     0.496799         0.485085   
min          0.000000     0.000000     0.000000     0.000000         0.000000   
25%          0.000000     0.000000     0.000000     0.000000         0.000000   
50%          1.000000     1.000000     0.000000     1.000000         1.000000   
75%          1.000000     1.000000     1.000000     1.000000         1.000000   
max          1.000000     1.000000     1.000000     1.000000         1.000000   

       teachers_guns   gun_safety  mental_health_screening  
count    1504.000000  1515.000000              1512.000000  
mean        0.294548     0.927393                 0.908069  
std         0.455991     0.259576                 0.289024  
min         0.000000     0.000000                 0.000000  
25%         0.000000     1.000000                 1.000000  
50%         0.000000     1.000000                 1.000000  
75%         1.000000     1.000000                 1.000000  
max         1.000000     1.000000                 1.000000

We wil move on to the independent variables. Each of them is either scale or categorical, thus some need to be converted to dummies in order to continue with the analysis. 2 groups are created, since we will perform regression with each of them as the paper did. Also, an attempt was made to have variable descriptives close with the paper's variables.

In [20]:
independent_variables1 = ["supernatural_evil", "attendance", "biblical_inerrancy", "biblical_literalism", \
    "mainline_protestant", "black_protestant", "conservative_protestant", "catholic", "other_religion"]
independent_variables2 = independent_variables1 + ["political_ideology", "age", "female", "non_hisp_black", "hispanic", "other_race", \
    "less_than_hs", "some_college", "college_graduate", "postgraduate", "income", "married", "children", \
    "urban", "south"]
independent_variables2

['supernatural_evil',
 'attendance',
 'biblical_inerrancy',
 'biblical_literalism',
 'mainline_protestant',
 'black_protestant',
 'conservative_protestant',
 'catholic',
 'other_religion',
 'political_ideology',
 'age',
 'female',
 'non_hisp_black',
 'hispanic',
 'other_race',
 'less_than_hs',
 'some_college',
 'college_graduate',
 'postgraduate',
 'income',
 'married',
 'children',
 'urban',
 'south']

Attendance does not need special handling, thus we take it as it is.

In [21]:
selected_vars['attendance'] = df["Q4"]
selected_vars['attendance'].describe()

count    1556.000000
mean        3.676735
std         2.932510
min         0.000000
25%         1.000000
50%         3.000000
75%         7.000000
max         8.000000
Name: attendance, dtype: float64

Bible errors and history-legends are treated as one value, thus we map them to 3, including the missing values.

In [22]:
df['Q17'] = df['Q17'].map({4: 3, 8: 3, 2: 1, 1: 1, 2: 2, 3: 3, np.nan: 3})
df['Q17'].value_counts() / df['Q17'].count()

3    0.440840
2    0.333333
1    0.225827
Name: Q17, dtype: float64

We then create the dummies. We do not use `dropFirst=True`, since we keep `biblical_literalism` variable.

In [23]:
df = pd.get_dummies(df, columns=["Q17"])
df.describe()

MOTHERLODE_ID    PROJECT_ID  PROJECT_NUMBER  TEMPLATE_NUMBER_162048140  \
count   1.572000e+03   1572.000000          1572.0                     1572.0   
mean    1.281718e+08  33370.021628     162048140.0                162048140.0   
std     5.894006e+02      0.145513             0.0                        0.0   
min     1.281709e+08  33370.000000     162048140.0                162048140.0   
25%     1.281713e+08  33370.000000     162048140.0                162048140.0   
50%     1.281717e+08  33370.000000     162048140.0                162048140.0   
75%     1.281722e+08  33370.000000     162048140.0                162048140.0   
max     1.281729e+08  33371.000000     162048140.0                162048140.0   

          ENTITY_ID  SCAN_RESPONDENT_ID  AUTO_INCREMENT       CASEID  \
count  1.572000e+03        1.572000e+03     1572.000000  1572.000000   
mean   4.139833e+09        4.139833e+09      831.400763   831.400763   
std    4.662307e+03        4.662307e+03      522.917758   522.917758   
min    4.139818e+09        4.139818e+09        1.000000     1.000000   
25%    4.139832e+09        4.139832e+09      376.750000   376.750000   
50%    4.139834e+09        4.139834e+09      792.500000   792.500000   
75%    4.139836e+09        4.139836e+09     1255.250000  1255.250000   
max    4.139838e+09        4.139838e+09     1864.000000  1864.000000   

          BATCH_ID           Q1  Q2_DK           Q3         Q3_1           Q4  \
count  1572.000000  1550.000000  254.0  1553.000000  1553.000000  1556.000000   
mean     33.784351    91.458710    1.0     2.816484     3.337411     3.676735   
std      20.930729   798.668391    0.0     1.221241     1.089783     2.932510   
min       1.000000     1.000000    1.0     1.000000     1.000000     0.000000   
25%      16.000000    16.000000    1.0     2.000000     3.000000     1.000000   
50%      32.000000    25.000000    1.0     3.000000     3.000000     3.000000   
75%      51.000000    45.000000    1.0     4.000000     4.000000     7.000000   
max      75.000000  9998.000000    1.0     8.000000     8.000000     8.000000   

                Q7           Q8          Q25          Q10          Q11  \
count  1143.000000  1147.000000  1549.000000  1551.000000  1540.000000   
mean      3.311461     3.154316     3.099419     2.900064     3.197403   
std       1.475759     2.170661     1.487391     1.832240     3.139593   
min       1.000000     1.000000     1.000000     0.000000     0.000000   
25%       2.000000     2.000000     2.000000     1.000000     0.000000   
50%       3.000000     2.000000     3.000000     3.000000     2.000000   
75%       5.000000     5.000000     3.000000     5.000000     6.000000   
max       5.000000     8.000000     8.000000     5.000000     8.000000   

              Q13A         Q13B         Q13C         Q13D         Q13E  \
count  1496.000000  1480.000000  1462.000000  1458.000000  1487.000000   
mean      1.500668     1.558784     1.184679     1.235254     1.717552   
std       0.802171     0.951454     0.582741     0.614572     0.964153   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     2.000000     1.000000     1.000000     2.000000   
max       4.000000     4.000000     4.000000     4.000000     4.000000   

              Q13F         Q13G         Q13H         Q13I         Q13J  \
count  1460.000000  1460.000000  1462.000000  1462.000000  1465.000000   
mean      1.376712     1.136301     1.092339     1.073187     1.096928   
std       0.739935     0.484068     0.407438     0.346261     0.367977   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     1.000000     1.000000  

And then keep then in helper DataFrame.

In [24]:
selected_vars[["biblical_inerrancy", "biblical_literalism"]] = df[["Q17_2", "Q17_1"]]
selected_vars[["biblical_inerrancy", "biblical_literalism"]].describe()

biblical_inerrancy  biblical_literalism
count         1572.000000          1572.000000
mean             0.333333             0.225827
std              0.471555             0.418259
min              0.000000             0.000000
25%              0.000000             0.000000
50%              0.000000             0.000000
75%              1.000000             0.000000
max              1.000000             1.000000

Now we will move to religion and gain the current percentages of values

In [25]:
df["RELTRAD"].value_counts() / df["RELTRAD"].count()

1.0    0.284267
4.0    0.249156
2.0    0.168805
7.0    0.144497
3.0    0.071573
6.0    0.059419
5.0    0.022282
Name: RELTRAD, dtype: float64

Jewish people and other religions should be grouped to match paper's descriptives.

In [26]:
df['RELTRAD'] = df['RELTRAD'].map({1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 5, 7: 7})
df["RELTRAD"].value_counts() / df["RELTRAD"].count()

1.0    0.284267
4.0    0.249156
2.0    0.168805
7.0    0.144497
5.0    0.081702
3.0    0.071573
Name: RELTRAD, dtype: float64

We then create dummies.

In [27]:
df = pd.get_dummies(df, columns=["RELTRAD"])
df.describe()

MOTHERLODE_ID    PROJECT_ID  PROJECT_NUMBER  TEMPLATE_NUMBER_162048140  \
count   1.572000e+03   1572.000000          1572.0                     1572.0   
mean    1.281718e+08  33370.021628     162048140.0                162048140.0   
std     5.894006e+02      0.145513             0.0                        0.0   
min     1.281709e+08  33370.000000     162048140.0                162048140.0   
25%     1.281713e+08  33370.000000     162048140.0                162048140.0   
50%     1.281717e+08  33370.000000     162048140.0                162048140.0   
75%     1.281722e+08  33370.000000     162048140.0                162048140.0   
max     1.281729e+08  33371.000000     162048140.0                162048140.0   

          ENTITY_ID  SCAN_RESPONDENT_ID  AUTO_INCREMENT       CASEID  \
count  1.572000e+03        1.572000e+03     1572.000000  1572.000000   
mean   4.139833e+09        4.139833e+09      831.400763   831.400763   
std    4.662307e+03        4.662307e+03      522.917758   522.917758   
min    4.139818e+09        4.139818e+09        1.000000     1.000000   
25%    4.139832e+09        4.139832e+09      376.750000   376.750000   
50%    4.139834e+09        4.139834e+09      792.500000   792.500000   
75%    4.139836e+09        4.139836e+09     1255.250000  1255.250000   
max    4.139838e+09        4.139838e+09     1864.000000  1864.000000   

          BATCH_ID           Q1  Q2_DK           Q3         Q3_1           Q4  \
count  1572.000000  1550.000000  254.0  1553.000000  1553.000000  1556.000000   
mean     33.784351    91.458710    1.0     2.816484     3.337411     3.676735   
std      20.930729   798.668391    0.0     1.221241     1.089783     2.932510   
min       1.000000     1.000000    1.0     1.000000     1.000000     0.000000   
25%      16.000000    16.000000    1.0     2.000000     3.000000     1.000000   
50%      32.000000    25.000000    1.0     3.000000     3.000000     3.000000   
75%      51.000000    45.000000    1.0     4.000000     4.000000     7.000000   
max      75.000000  9998.000000    1.0     8.000000     8.000000     8.000000   

                Q7           Q8          Q25          Q10          Q11  \
count  1143.000000  1147.000000  1549.000000  1551.000000  1540.000000   
mean      3.311461     3.154316     3.099419     2.900064     3.197403   
std       1.475759     2.170661     1.487391     1.832240     3.139593   
min       1.000000     1.000000     1.000000     0.000000     0.000000   
25%       2.000000     2.000000     2.000000     1.000000     0.000000   
50%       3.000000     2.000000     3.000000     3.000000     2.000000   
75%       5.000000     5.000000     3.000000     5.000000     6.000000   
max       5.000000     8.000000     8.000000     5.000000     8.000000   

              Q13A         Q13B         Q13C         Q13D         Q13E  \
count  1496.000000  1480.000000  1462.000000  1458.000000  1487.000000   
mean      1.500668     1.558784     1.184679     1.235254     1.717552   
std       0.802171     0.951454     0.582741     0.614572     0.964153   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     2.000000     1.000000     1.000000     2.000000   
max       4.000000     4.000000     4.000000     4.000000     4.000000   

              Q13F         Q13G         Q13H         Q13I         Q13J  \
count  1460.000000  1460.000000  1462.000000  1462.000000  1465.000000   
mean      1.376712     1.136301     1.092339     1.073187     1.096928   
std       0.739935     0.484068     0.407438     0.346261     0.367977   
min       1.000000     1.000000     1.000000     1.000000     1.000000   
25%       1.000000     1.000000     1.000000     1.000000     1.000000   
50%       1.000000     1.000000     1.000000     1.000000     1.000000   
75%       2.000000     1.000000     1.000000  

We do not keep no affiliation.

In [28]:
selected_vars[["mainline_protestant", "black_protestant", \
    "conservative_protestant", "catholic", "other_religion"]] = df[["RELTRAD_2.0", \
    "RELTRAD_3.0", "RELTRAD_1.0", "RELTRAD_4.0", "RELTRAD_5.0"]]
selected_vars[["mainline_protestant", "black_protestant", \
    "conservative_protestant", "catholic", "other_religion"]].describe()

mainline_protestant  black_protestant  conservative_protestant  \
count          1572.000000       1572.000000              1572.000000   
mean              0.159033          0.067430                 0.267812   
std               0.365823          0.250845                 0.442960   
min               0.000000          0.000000                 0.000000   
25%               0.000000          0.000000                 0.000000   
50%               0.000000          0.000000                 0.000000   
75%               0.000000          0.000000                 1.000000   
max               1.000000          1.000000                 1.000000   

          catholic  other_religion  
count  1572.000000     1572.000000  
mean      0.234733        0.076972  
std       0.423967        0.266632  
min       0.000000        0.000000  
25%       0.000000        0.000000  
50%       0.000000        0.000000  
75%       0.000000        0.000000  
max       1.000000        1.000000

We move to political ideology.

In [29]:
df['Q31'].describe()

count    1521.000000
mean        3.800789
std         1.554416
min         1.000000
25%         2.000000
50%         4.000000
75%         5.000000
max         7.000000
Name: Q31, dtype: float64

It is close so we will pass it directly to the other dataframe.

In [30]:
selected_vars['political_ideology'] = df['Q31']
selected_vars.political_ideology

0       1.0
1       4.0
2       6.0
3       2.0
4       2.0
       ... 
1567    3.0
1568    4.0
1569    4.0
1570    6.0
1571    NaN
Name: political_ideology, Length: 1572, dtype: float64

In [31]:
df["AGE"].describe()

count    1572.000000
mean       49.162214
std        21.018613
min         0.000000
25%        35.000000
50%        52.000000
75%        64.000000
max        99.000000
Name: AGE, dtype: float64

We will do the same for age.

In [32]:
selected_vars['age'] = df['AGE']
selected_vars.age

0       85
1        0
2       67
3       64
4       54
        ..
1567    48
1568    59
1569    29
1570    57
1571     0
Name: age, Length: 1572, dtype: int64

For gender, we only neet to map 1 & 2 to 0 (male) & 1 (female).

In [33]:
df['Q77'] = df['Q77'].map({2: 1, 1: 0})
df["Q77"].value_counts() / df["Q77"].count()

1.0    0.578776
0.0    0.421224
Name: Q77, dtype: float64

Dummy is ready, so we will pass it to `selected_vars`.

In [34]:
selected_vars["female"] = df["Q77"]
selected_vars["female"].describe()

count    1536.000000
mean        0.578776
std         0.493916
min         0.000000
25%         0.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: female, dtype: float64

Race emerges from several questions. We will encode whites to see if descriptives agree with the paper, although they will not be further used in the analysis.

In [35]:
mapping = {
    1: 1,
    2: 0,
    np.nan: 0
}
df["Q88A"] = df["Q88A"].map(mapping) 
df["Q88A"].value_counts() / df["Q88A"].count()

1    0.746183
0    0.253817
Name: Q88A, dtype: float64

Continuing with blacks.

In [36]:
df["Q88B"] = df["Q88B"].map(mapping)
selected_vars['non_hisp_black'] = df["Q88B"]
df["Q88B"].value_counts() / df["Q88B"].count()

0    0.886132
1    0.113868
Name: Q88B, dtype: float64

And for hispanics, we will map specific hispanic races to hispanic.

In [37]:
df["Q89"] = df["Q89"].map({1: 0, 2: 1, 3: 1, 4: 1, 5: 1, 8: 1, np.nan: 0})
selected_vars['hispanic'] = df["Q89"]
df["Q89"].value_counts() / df["Q89"].count()

0    0.877863
1    0.122137
Name: Q89, dtype: float64

Others will be calculated as people who do not belong to the previous races.

In [38]:
df.loc[(df["Q88A"] == 0) & (df["Q88B"] == 0) & (df["Q89"] == 0), "other"] = 1
df.loc[(df["Q88A"] == 1) | (df["Q88B"] == 1) | (df["Q89"] == 1), "other"] = 0
selected_vars['other_race'] = df["other"]
selected_vars.describe()

supernatural_evil  semi_auto_guns   ammo_clips     handguns  \
count        1517.000000     1502.000000  1502.000000  1501.000000   
mean            3.089541        0.629161     0.671105     0.290473   
std             1.053424        0.483190     0.469968     0.454132   
min             1.000000        0.000000     0.000000     0.000000   
25%             2.333333        0.000000     0.000000     0.000000   
50%             3.666667        1.000000     1.000000     0.000000   
75%             4.000000        1.000000     1.000000     1.000000   
max             4.000000        1.000000     1.000000     1.000000   

        carry_laws  school_security  teachers_guns   gun_safety  \
count  1502.000000      1510.000000    1504.000000  1515.000000   
mean      0.557923         0.621854       0.294548     0.927393   
std       0.496799         0.485085       0.455991     0.259576   
min       0.000000         0.000000       0.000000     0.000000   
25%       0.000000         0.000000       0.000000     1.000000   
50%       1.000000         1.000000       0.000000     1.000000   
75%       1.000000         1.000000       1.000000     1.000000   
max       1.000000         1.000000       1.000000     1.000000   

       mental_health_screening   attendance  biblical_inerrancy  \
count              1512.000000  1556.000000         1572.000000   
mean                  0.908069     3.676735            0.333333   
std                   0.289024     2.932510            0.471555   
min                   0.000000     0.000000            0.000000   
25%                   1.000000     1.000000            0.000000   
50%                   1.000000     3.000000            0.000000   
75%                   1.000000     7.000000            1.000000   
max                   1.000000     8.000000            1.000000   

       biblical_literalism  mainline_protestant  black_protestant  \
count          1572.000000          1572.000000       1572.000000   
mean              0.225827             0.159033          0.067430   
std               0.418259             0.365823          0.250845   
min               0.000000             0.000000          0.000000   
25%               0.000000             0.000000          0.000000   
50%               0.000000             0.000000          0.000000   
75%               0.000000             0.000000          0.000000   
max               1.000000             1.000000          1.000000   

       conservative_protestant     catholic  other_religion  \
count              1572.000000  1572.000000     1572.000000   
mean                  0.267812     0.234733        0.076972   
std                   0.442960     0.423967        0.266632   
min                   0.000000     0.000000        0.000000   
25%                   0.000000     0.000000        0.000000   
50%                   0.000000     0.000000        0.000000   
75%                   1.000000     0.000000        0.000000   
max                   1.000000     1.000000        1.000000   

       political_ideology          age       female  non_hisp_black  \
count         1521.000000  1572.000000  1536.000000     1572.000000   
mean             3.800789    49.162214     0.578776        0.113868   
std              1.554416    21.018613     0.493916        0.317752   
min              1.000000     0.000000     0.000000        0.000000   
25%              2.000000    35.000000     0.000000        0.000000   
50%              4.000000    52.000000     1.000000        0.000000   
75%              5.000000    64.000000     1.000000        0.000000   
max              7.000000    99.000000     1.000000        1.000000   

          hispanic   other_race  
count  1572.000000  1572.000000  
mean      0.122137     0.087150  
std       0.327549     0.282145  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000     1.000000

Moving on to education, _Trade/Technical/Vocational training_ will be mapped together with _Some college_, since this results in closer to the paper's descriptives. 

In [39]:
df["I_EDUC"] = df["I_EDUC"].map({
                            1: 1,
                            2: 2,
                            3: 3,
                            4: 3,
                            5: 5,
                            6: 6
                        })

df["I_EDUC"].value_counts() / df["I_EDUC"].count()


3.0    0.321785
5.0    0.261825
6.0    0.223185
2.0    0.133911
1.0    0.059294
Name: I_EDUC, dtype: float64

We will now create dummies

In [40]:
df = pd.get_dummies(df, columns=["I_EDUC"])
df

MOTHERLODE_ID  PROJECT_ID       RESPONDENT_ID     METHOD_ID METHOD_TYPE  \
0         128170892       33370  01_000001_00000001  162048140_01        SCAN   
1         128170893       33370  01_000002_00000001  162048140_01        SCAN   
2         128170894       33370  01_000003_00000001  162048140_01        SCAN   
3         128170895       33370  01_000004_00000001  162048140_01        SCAN   
4         128170896       33370  01_000005_00000001  162048140_01        SCAN   
...             ...         ...                 ...           ...         ...   
1567      128172920       33371  02_000301_00000013  162048140_02        SCAN   
1568      128172921       33371  02_000326_00000014  162048140_02        SCAN   
1569      128172922       33371  02_000351_00000015  162048140_02        SCAN   
1570      128172335       33371  02_000352_00000015  162048140_02        SCAN   
1571      128172923       33371  02_000376_00000016  162048140_02        SCAN   

      PROJECT_NUMBER CREATED_ON PRACTICE RESPONDENT_DATE RESPONDENT_LANGUAGE  \
0          162048140  3/26/2014    OTHER       1/24/2014               en-US   
1          162048140  3/26/2014    OTHER       1/24/2014               en-US   
2          162048140  3/26/2014    OTHER       1/24/2014               en-US   
3          162048140  3/26/2014    OTHER       1/24/2014               en-US   
4          162048140  3/26/2014    OTHER       1/24/2014               en-US   
...              ...        ...      ...             ...                 ...   
1567       162048140  3/26/2014    OTHER       3/10/2014               es-AR   
1568       162048140  3/26/2014    OTHER       3/12/2014               es-AR   
1569       162048140  3/26/2014    OTHER       3/19/2014               es-AR   
1570       162048140  3/26/2014    OTHER       3/19/2014               es-AR   
1571       162048140  3/26/2014    OTHER       3/21/2014               es-AR   

      TEMPLATE_NUMBER_162048140   ENTITY_ID  SCAN_RESPONDENT_ID  \
0                     162048140  4139835441          4139835441   
1                     162048140  4139832983          4139832983   
2                     162048140  4139835115          4139835115   
3                     162048140  4139837428          4139837428   
4                     162048140  4139835007          4139835007   
...                         ...         ...                 ...   
1567                  162048140  4139818420          4139818420   
1568                  162048140  4139818965          4139818965   
1569                  162048140  4139819569          4139819569   
1570                  162048140  4139830483          4139830483   
1571                  162048140  4139830602          4139830602   

      AUTO_INCREMENT  CASEID  BATCH_ID  LANG1    Q1  Q2_DK   Q3  Q3_1   Q4  \
0                  1       1         1  en-US  20.0    NaN  4.0   4.0  8.0   
1                  2       2         1  en-US   NaN    NaN  4.0   4.0  1.0   
2                  3       3         1  en-US  31.0    NaN  3.0   4.0  7.0   
3                  4       4         1  en-US  33.0    NaN  3.0   4.0  6.0   
4                  5       5         1  en-US  14.0    NaN  4.0   4.0  8.0   
...              ...     ...       ...    ...   ...    ...  ...   ...  ...   
1567             301     301        13  es-AR  16.0    NaN  3.0   NaN  1.0   
1568             326     326        14  es-AR  16.0    NaN  8.0   3.0  0.0   
1569             351     351        15  es-AR   1.0    NaN  4.0   3.0  7.0   
1570             352     352        15  es-AR  16.0    NaN  4.0   4.0  3.0   
1571             376     376        16  es-AR  16.0    1.0  3.0   3.0  5.0   

       Q7   Q8  Q25  Q10  Q11  Q13A  Q13B  Q13C  Q13D  Q13E  Q13F  Q13G  Q13H  \
0     5.0  1.0  3.0  5.0  8.0   2.0   4.0   1.0   1.0   2.0   1.0   1.0   1.0   
1     1.0  8.0  3.0  3.0  5.0   1.0   1.0   1.0   2.0   2.0   2.0   1.0   1.0   
2     5.0  5.0  3.0  5.0  1.0   2.0   1.0   1.0   1.0   3.0   2.0   1.0   1.0   
3     

We will ommit column `I_EDUC_2.0` (high school graduate).

In [41]:
selected_vars[["less_than_hs", "some_college", "college_graduate", "postgraduate"]] = df[["I_EDUC_1.0", "I_EDUC_3.0", "I_EDUC_5.0", "I_EDUC_6.0"]]
selected_vars.describe()

supernatural_evil  semi_auto_guns   ammo_clips     handguns  \
count        1517.000000     1502.000000  1502.000000  1501.000000   
mean            3.089541        0.629161     0.671105     0.290473   
std             1.053424        0.483190     0.469968     0.454132   
min             1.000000        0.000000     0.000000     0.000000   
25%             2.333333        0.000000     0.000000     0.000000   
50%             3.666667        1.000000     1.000000     0.000000   
75%             4.000000        1.000000     1.000000     1.000000   
max             4.000000        1.000000     1.000000     1.000000   

        carry_laws  school_security  teachers_guns   gun_safety  \
count  1502.000000      1510.000000    1504.000000  1515.000000   
mean      0.557923         0.621854       0.294548     0.927393   
std       0.496799         0.485085       0.455991     0.259576   
min       0.000000         0.000000       0.000000     0.000000   
25%       0.000000         0.000000       0.000000     1.000000   
50%       1.000000         1.000000       0.000000     1.000000   
75%       1.000000         1.000000       1.000000     1.000000   
max       1.000000         1.000000       1.000000     1.000000   

       mental_health_screening   attendance  biblical_inerrancy  \
count              1512.000000  1556.000000         1572.000000   
mean                  0.908069     3.676735            0.333333   
std                   0.289024     2.932510            0.471555   
min                   0.000000     0.000000            0.000000   
25%                   1.000000     1.000000            0.000000   
50%                   1.000000     3.000000            0.000000   
75%                   1.000000     7.000000            1.000000   
max                   1.000000     8.000000            1.000000   

       biblical_literalism  mainline_protestant  black_protestant  \
count          1572.000000          1572.000000       1572.000000   
mean              0.225827             0.159033          0.067430   
std               0.418259             0.365823          0.250845   
min               0.000000             0.000000          0.000000   
25%               0.000000             0.000000          0.000000   
50%               0.000000             0.000000          0.000000   
75%               0.000000             0.000000          0.000000   
max               1.000000             1.000000          1.000000   

       conservative_protestant     catholic  other_religion  \
count              1572.000000  1572.000000     1572.000000   
mean                  0.267812     0.234733        0.076972   
std                   0.442960     0.423967        0.266632   
min                   0.000000     0.000000        0.000000   
25%                   0.000000     0.000000        0.000000   
50%                   0.000000     0.000000        0.000000   
75%                   1.000000     0.000000        0.000000   
max                   1.000000     1.000000        1.000000   

       political_ideology          age       female  non_hisp_black  \
count         1521.000000  1572.000000  1536.000000     1572.000000   
mean             3.800789    49.162214     0.578776        0.113868   
std              1.554416    21.018613     0.493916        0.317752   
min              1.000000     0.000000     0.000000        0.000000   
25%              2.000000    35.000000     0.000000        0.000000   
50%              4.000000    52.000000     1.000000        0.000000   
75%              5.000000    64.000000     1.000000        0.000000   
max              7.000000    99.000000     1.000000        1.000000   

          hispanic   other_race  less_than_hs  some_college  college_graduate  \
count  1572.000000  1572.000000   1572.000000   1572.000000        1572.00000   
mean      0.122137     0.087150      0.056616      0.307252           0.25000   
std       0.327549     0.282145      0.231181      0.461502           0.43315   
min       0.00

Income does not need any processing.

In [42]:
selected_vars["income"] = df['Q95']
selected_vars['income'].describe()

count    1467.000000
mean        4.151329
std         1.707939
min         1.000000
25%         3.000000
50%         5.000000
75%         5.000000
max         7.000000
Name: income, dtype: float64

Moving on to marriage.

In [43]:
df['Q51A'].value_counts() / df['Q51A'].count()

2.0    0.513514
1.0    0.198842
5.0    0.139640
6.0    0.084942
3.0    0.039254
4.0    0.023810
Name: Q51A, dtype: float64

_Married_ and _Living as married/partnered_ will only be considered as married.

In [44]:
mapping = {
    1: 0,
    2: 1,
    3: 1,
    4: 0,
    5: 0,
    6: 0
}
selected_vars['married'] = df['Q51A'].map(mapping)
selected_vars['married'].describe()

count    1554.000000
mean        0.552767
std         0.497368
min         0.000000
25%         0.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: married, dtype: float64

We are only interested in whether or not exist children at home and not in how many.

In [45]:
selected_vars['children'] = df['Q93'].apply(lambda x: 1 if x > 0 else 0)
selected_vars.children.describe()

count    1572.000000
mean        0.254453
std         0.435692
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: children, dtype: float64

Cotinuing with region.

In [46]:
df['Q80'].value_counts() / df['Q80'].count()

3.0    0.330519
2.0    0.270130
1.0    0.250649
4.0    0.135714
8.0    0.012987
Name: Q80, dtype: float64

Only _A large city_ and _A suburb near a large city_ are mapped as urban.

In [47]:
mapping = {
    1: 1,
    2: 1,
    3: 0,
    4: 0,
    8: 0
}
selected_vars['urban'] = df['Q80'].map(mapping)

Finally, we have an abbreviation for the state where people live.

In [48]:
df['STATE']

0       TX
1       TX
2       ND
3       TX
4       IA
        ..
1567    CA
1568    CA
1569    CA
1570    CT
1571    GA
Name: STATE, Length: 1572, dtype: object

Using the following links, we find and map the southern states.
* https://www.bls.gov/respondents/mwr/electronic-data-interchange/appendix-d-usps-state-abbreviations-and-fips-codes.htm
* https://en.wikipedia.org/wiki/Southern_United_States

In [49]:
southern_states = ['TX', 'FL', 'GA', 'NC', 'VA', 'TN', 'MD', 'SC', 'AL', 'LA', \
    'KY', 'OK', 'AR', 'MS', 'WV', 'DE']
selected_vars['south'] = df['STATE'].apply(lambda x: 1 if x in southern_states else 0)
selected_vars['south'].describe()

count    1572.000000
mean        0.382952
std         0.486261
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Name: south, dtype: float64

The variables that will be used are the following:

In [50]:
selected_vars.describe()

supernatural_evil  semi_auto_guns   ammo_clips     handguns  \
count        1517.000000     1502.000000  1502.000000  1501.000000   
mean            3.089541        0.629161     0.671105     0.290473   
std             1.053424        0.483190     0.469968     0.454132   
min             1.000000        0.000000     0.000000     0.000000   
25%             2.333333        0.000000     0.000000     0.000000   
50%             3.666667        1.000000     1.000000     0.000000   
75%             4.000000        1.000000     1.000000     1.000000   
max             4.000000        1.000000     1.000000     1.000000   

        carry_laws  school_security  teachers_guns   gun_safety  \
count  1502.000000      1510.000000    1504.000000  1515.000000   
mean      0.557923         0.621854       0.294548     0.927393   
std       0.496799         0.485085       0.455991     0.259576   
min       0.000000         0.000000       0.000000     0.000000   
25%       0.000000         0.000000       0.000000     1.000000   
50%       1.000000         1.000000       0.000000     1.000000   
75%       1.000000         1.000000       1.000000     1.000000   
max       1.000000         1.000000       1.000000     1.000000   

       mental_health_screening   attendance  biblical_inerrancy  \
count              1512.000000  1556.000000         1572.000000   
mean                  0.908069     3.676735            0.333333   
std                   0.289024     2.932510            0.471555   
min                   0.000000     0.000000            0.000000   
25%                   1.000000     1.000000            0.000000   
50%                   1.000000     3.000000            0.000000   
75%                   1.000000     7.000000            1.000000   
max                   1.000000     8.000000            1.000000   

       biblical_literalism  mainline_protestant  black_protestant  \
count          1572.000000          1572.000000       1572.000000   
mean              0.225827             0.159033          0.067430   
std               0.418259             0.365823          0.250845   
min               0.000000             0.000000          0.000000   
25%               0.000000             0.000000          0.000000   
50%               0.000000             0.000000          0.000000   
75%               0.000000             0.000000          0.000000   
max               1.000000             1.000000          1.000000   

       conservative_protestant     catholic  other_religion  \
count              1572.000000  1572.000000     1572.000000   
mean                  0.267812     0.234733        0.076972   
std                   0.442960     0.423967        0.266632   
min                   0.000000     0.000000        0.000000   
25%                   0.000000     0.000000        0.000000   
50%                   0.000000     0.000000        0.000000   
75%                   1.000000     0.000000        0.000000   
max                   1.000000     1.000000        1.000000   

       political_ideology          age       female  non_hisp_black  \
count         1521.000000  1572.000000  1536.000000     1572.000000   
mean             3.800789    49.162214     0.578776        0.113868   
std              1.554416    21.018613     0.493916        0.317752   
min              1.000000     0.000000     0.000000        0.000000   
25%              2.000000    35.000000     0.000000        0.000000   
50%              4.000000    52.000000     1.000000        0.000000   
75%              5.000000    64.000000     1.000000        0.000000   
max              7.000000    99.000000     1.000000        1.000000   

          hispanic   other_race  less_than_hs  some_college  college_graduate  \
count  1572.000000  1572.000000   1572.000000   1572.000000        1572.00000   
mean      0.122137     0.087150      0.056616      0.307252           0.25000   
std       0.327549     0.282145      0.231181      0.461502           0.43315   
min       0.00

### Predict Support for Various Gun Policies

Logistic Regressions

In [ ]:
def log_regress(independent_variables):
    all_columns_formula = "+".join(independent_variables)
    for var in dependent_variables:
        model = logit(var + " ~ " + all_columns_formula, selected_vars, ).fit()
        print(model.summary2())
        coefficients = dict(model.params)
        print("ODD RATIOS")
        for k in coefficients.keys():
            print('{0:10}  {1}'.format(k, np.exp(coefficients[k])))

In [ ]:
log_regress(independent_variables1)

### Results

#### Semi-Auto Weapons Ban

Statistically important variables (p-value < 0.05) are `supernatural_evil`, `black_protestant`, `catholic` and `other_religion`. This means that:
* One unit increase in the supernatural evil metric brings a decrease of 0.32 in the log odds of supporting semi-auto weapons ban; equivalently, the odds decrease by 1 - 0.73 = 27%.
* The fact that someone is black protestant brings an increase of 0.66 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 93%.
* The fact that someone is catholic brings an increase of 0.4 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 49%.
* The fact that someone is of other religion brings an increase of 0.7 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 100%.

#### Ban on High-Capacity Ammo Clips
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `black_protestant` and `other_religion`. This means that:
* One unit increase in the supernatural evil metric brings a decrease of 0.3 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds decrease by 1 - 0.74 = 26%.
* The fact that someone is black protestant brings an increase of 0.68 in the log odds of supporting high-capacity ammo clipss ban; equivalently, the odds increase by 97%.
* The fact that someone is of other religion brings an increase of 0.64 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 90%.

#### Banning Civilian Handguns
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `attendance`,`black_protestant` and `other_religion`. This means that:
* One unit increase in the supernatural evil metric brings a decrease of 0.42 in the log odds of supporting civilian handguns ban; equivalently, the odds decrease by 1 - 0.66 = 34%.
* One unit increase in attendance brings an increase of 0.06 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 6%.
* The fact that someone is black protestant brings an increase of 1.1 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 200.1%.
* The fact that someone is of other religion brings an increase of 0.53 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 70%.

#### Support for Concealed Carry Laws 
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `black_protestant`, `catholic` and `other_religion`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.42 in the log odds of supporting concealed carry laws; equivalently, the odds increase by 52%.
* The fact that someone is black protestant brings a decrease of 0.94 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.39 = 61%.
* The fact that someone is catholic brings a decrease of 0.42 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.65 = 35%.
* The fact that someone is of other religion brings a decrease of 0.62 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.54 = 46%.

#### More Armed Security at Schools 
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `attendance` and `biblical_literalism`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.46 in the log odds of supporting armed security at schools; equivalently, the odds increase by 58%.
* One unit increase in attendance brings a decrease of 0.08 in the log odds of supporting armed security at schools; equivalently, the odds decrease by 1 - 0.92 = 8%.
* The fact that someone believes in biblical literalism brings an increase of 0.79 in the log odds of supporting armed security at schools; equivalently, the odds increase by 119%.

#### More Teachers/Faculty having Guns
The only statistically important variable (p-value < 0.05) is `supernatural_evil`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.38 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds increase by 46%.

#### More Gun Safety Programs 
No statistically important variables were found.

#### Expanded Mental Health Screening
No statistically important variables were found.

We will repeat regression for all variables.

In [ ]:
log_regress(independent_variables2)

### Results
By itroducing new variables in the model, the previous coefficients decrease, as part of the result is explained by others.

#### Semi-Auto Weapons Ban
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `conservative_protestant`, `catholic`, `political_ideology`, `age`, `female`, `income`, `urban` and `south`. This means that:
* One unit increase in the supernatural evil metric brings a decrease of 0.21 in the log odds of supporting semi-auto weapons ban; equivalently, the odds decrease by 1 - 0.81 = 19%.
* The fact that someone is conservative protestant brings an increase of 0.58 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 78%.
* The fact that someone is catholic brings an increase of 0.58 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 78%. (There is a conflict with the paper here.)
* One unit increase in political liberalism brings an increase of 0.52 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 69%.
* One unit increase in age brings an increase of 0.2 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 1%.
* The fact that someone is female brings an increase of 1.2 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 231%.
* One unit increase in income brings an increase of 0.11 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 12%.
* The fact that someone lives in an urban area brings an increase of 0.35 in the log odds of supporting semi-auto weapons ban; equivalently, the odds increase by 42%.
* The fact that someone lives in a southern state brings a decrease of 0.31 in the log odds of supporting semi-auto weapons ban; equivalently, the odds decrease by 1 - 0.73 = 27%.

#### Ban on High-Capacity Ammo Clips
Statistically important variables (p-value < 0.05) are `political_ideology`, `age`, `female`, `postgraduate` and `urban`. This means that:
* One unit increase in political liberalism brings an increase of 0.47 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 60%.
* One unit increase in age brings an increase of 0.2 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 2%.
* The fact that someone is female brings an increase of 1.03 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 180%.
* The fact that someone is postgraduate brings an increase of 0.54 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 71%. This correlation could be omitted, since its p-value is relatively high (0.0431). 
* The fact that someone lives in an urban area brings an increase of 0.33 in the log odds of supporting high-capacity ammo clips ban; equivalently, the odds increase by 39%.

#### Cilivian Hand Gun Possession Ban
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `attendance`, `political_ideology`, `hispanic`, `other_race`, `urban` and `south`. This means that:
* One unit increase in the supernatural evil metric brings a decrease of 0.29 in the log odds of supporting civilian handguns ban; equivalently, the odds decrease by 1 - 0.75 = 25%.
* One unit increase in attendance brings an increase of 0.07 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 8%.
* One unit increase in political liberalism brings an increase of 0.34 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 41%.
* The fact that someone is hispanic brings an increase of 0.99 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 169%.
* The fact that someone is of other race brings an increase of 1.16 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 219%.
* The fact that someone lives in an urban area brings an increase of 0.29 in the log odds of supporting civilian handguns ban; equivalently, the odds increase by 34%.
* The fact that someone lives in a southern state brings a decrease of 0.44 in the log odds of supporting civilian handguns ban; equivalently, the odds decrease by 1 - 0.65 = 35%.

Age is also a significant variable in the papepr, however there is not enough support in this particular analysis.

#### Support for Concealed Carry Laws
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `biblical_inerrancy`, `black_protestant`, `catholic`, `political_ideology`, `female`, `other_race`, `urban` and `south`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.32 in the log odds of supporting concealed carry laws; equivalently, the odds increase by 38%.
* The fact that someone believes in biblical inerrancy brings a decrease of 0.41 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.66 = 34%.
* The fact that someone is black protestant brings a decrease of 0.94 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.39 = 61%.
* The fact that someone is catholic brings a decrease of 0.46 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.63 = 37%.
* The fact that someone is catholic brings a decrease of 0.46 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.63 = 37%.
* One unit increase in political liberalism brings a decrease of 0.41 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.66 = 34%.
* The fact that someone is female brings a decrease of 0.68 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.5 = 50%.
* The fact that someone is of other race brings a decrease of 0.65 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.52 = 48%.
* The fact that someone lives in an urban area brings a decrease of 0.3 in the log odds of supporting concealed carry laws; equivalently, the odds decrease by 1 - 0.74 = 26%.
* The fact that someone lives in a southern state brings an increase of 0.43 in the log odds of supporting concealed carry laws; equivalently, the odds increase by 54%.

What is different, compared to the paper, is that the particular analysis found more religious variables as significant.

#### More Armed Security at Schools 
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `attendance`, `political_ideology`, `postgraduate`, `income`, `married` and `south`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.39 in the log odds of supporting armed security at schools; equivalently, the odds increase by 47%.
* One unit increase in attendance brings a decrease of 0.1 in the log odds of supporting armed security at schools; equivalently, the odds decrease by 1 - 0.9 = 10%.
* One unit increase in political liberalism brings a decrease of 0.23 in the log odds of supporting armed security at schools; equivalently, the odds decrease by 1 - 0.79 = 21%.
* The fact that someone is postgraduate brings a decrease of 0.62 in the log odds of supporting armed security at schools; equivalently, the odds decrease by 1 - 0.53 = 47%.
* One unit increase in income brings a decrease of 0.15 in the log odds of supporting armed security at schools; equivalently, the odds decrease by 1 - 0.86 = 14%.
* The fact that someone is married or cohabitating brings an increase of 0.29 in the log odds of supporting armed security at schools; equivalently, the odds increase by 34%.
* The fact that someone lives in a southern state brings an increase of 0.4 in the log odds of supporting armed security at schools; equivalently, the odds increase by 49%.

#### More Teachers/Faculty having Guns
Statistically important variables (p-value < 0.05) are `supernatural_evil`, `attendance`, `biblical_literalism`, `political_ideology`, `female`, `non_hisp_black`, `postgraduate` and `urban`. This means that:
* One unit increase in the supernatural evil metric brings an increase of 0.33 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds increase by 40%.
* One unit increase in attendance brings a decrease of 0.06 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.94 = 6%.
* The fact that someone believes in biblical literalism brings a decrease of 0.46 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.63 = 37%.
* One unit increase in political liberalism brings a decrease of 0.4 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.67 = 33%.
* The fact that someone is female brings a decrease of 0.72 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.49 = 51%.
* The fact that someone is black brings a decrease of 0.89 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.41 = 59%.
* The fact that someone is postgraduate brings a decrease of 0.53 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.59 = 41%.
* The fact that someone lives in an urban area brings a decrease of 0.46 in the log odds of supporting teachers or faculty carrying guns; equivalently, the odds decrease by 1 - 0.63 = 37%.

#### Required Gun Safety Programs
Statistically important variables (p-value < 0.05) are `political_ideology`, `age`, `female` and `urban`.  This means that:
* One unit increase in political liberalism brings an increase of 0.38 in the log odds of supporting gun safety programs; equivalently, the odds increase by 46%.
* One unit increase in age brings an increase of 0.01 in the log odds of supporting gun safety programs; equivalently, the odds increase by 1%.
* The fact that someone is female brings an increase of 0.62 in the log odds of supporting gun safety programs; equivalently, the odds increase by 86%.
* The fact that someone lives in an urban area brings an increase of 0.53 in the log odds of supporting gun safety programs; equivalently, the odds increase by 70%.

#### Expanded Mental Health Screening
Statistically important variables (p-value < 0.05) are `political_ideology`, `age`, `female`, `hispanic` and `children`. This means that:
* One unit increase in political liberalism brings an increase of 0.43 in the log odds of supporting expanded screening of mental health; equivalently, the odds increase by 53%.
* One unit increase in age brings an increase of 0.01 in the log odds of supporting expanded screening of mental health; equivalently, the odds increase by 1%.
* The fact that someone is female brings an increase of 0.74 in the log odds of supporting expanded screening of mental health; equivalently, the odds increase by 101%.
* The fact that someone is hispanic brings an increase of 1.13 in the log odds of supporting expanded screening of mental health; equivalently, the odds increase by 211%.
* The fact that someone has underaged children at home brings a decrease of 0.48 in the log odds of supporting expanded screening of mental health; equivalently, the odds increase by 1 - 0.62 = 38%.

### Economist Article

The article presents some inaccuracies. First, the supernatural evil metric was not found statistically important for all 8 dependent variables. (_"belief in the devil, demons and hell is a strong predictor of eight pro-gun beliefs, including arming teachers, carrying concealed firearms and bearing high-capacity defensive weapons."_). Furthermore, the analysis does not examine the correlation between independent variables. Thus, the statement _"A Catholic who believes in supernatural evil is more likely to hold pro-gun views than a Protestant who does not believe that Satan is corrupting souls, and vice versa."_ is inaccurate, since we have not examined whether or not catholics and protestants believe in supernatural evil.

### Additional Estimations of the Strength of Predictors

In order to 

In [ ]:
from sklearn.impute import SimpleImputer
scaled_selected = pd.DataFrame(selected_vars)

imp = SimpleImputer(strategy='mean')
scaled_selected[selected_vars.columns] = imp.fit_transform(selected_vars[selected_vars.columns])
scaled_selected

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaled_selected[selected_vars.columns] = scaler.fit_transform(selected_vars[selected_vars.columns])
scaled_selected

In [ ]:
def regress(independent_variables):
    all_columns_formula = "+".join(independent_variables)
    for var in dependent_variables:
        model = smf.ols(var + " ~ " + all_columns_formula, selected_vars)
        res = model.fit()
        print(res.summary())

In [ ]:
regress(independent_variables1)

Unfortunately, adjusted $$R^2$$s are too low, however we will make an attempt to interprete the results. Also, it is clear from the intercept, that much has to be explained from other variables that have been omitted.

#### Semi-Auto Weapons Ban
As in logistic regression (except from `catholic`), important variables are `supernatural_evil`, `black_protestant` and `other_religion`. Being a black protestant or of another religion, contributes positively to believing that a basn should be established. On the other hand, supernatural evil affects its believers negatively.

#### Ban on High-Capacity Ammo Clips
Again important variables are `supernatural_evil`, `black_protestant` and `other_religion`. The results are similar with the previous variable. 

#### Banning Civilian Handguns
Important variables are `supernatural_evil`, `catholic`, `black_protestant` and `other_religion`. It seems that being black protestant influences more positive this variable compared to the previous ones (0.2362 > 0.1420 and 0.1294). The intercept is lower and this may imply that religious variables explain in a higher degree attitude toward civilian handguns banning.

#### Support for Concealed Carry Laws 
Here, religious people, except from conservative protestants, seem to be affected negatively. Again, black protestants have the strongest influence compared to others. However, belief in supernatural evil affects people positively.

#### More Armed Security at Schools 
Statistically important variables are `supernatural_evil`, `attendance` and `biblical_literalism`, as in logistic regression. People believing in supernatural evil or interpreting literally what Bible says seem to believe more that there should ba armed security at schools. Attending religious service influences negative this attitude.

#### More Teachers/Faculty having Guns
Here `biblical_literalism` is not important, however `supernatural_evil` and `attendance` affect similarly with the previous variable the attiude towards armed teachers and faculty.

#### More Gun Safety Programs 
Again no statistically important variables were found.

#### Expanded Mental Health Screening
Again no statistically important variables were found.

We will repeat linear regression for all the variables.

In [ ]:
regress(independent_variables2)

Adjusted $$R^s$$ are better now since we added more variables. We will point out some interesting results:

#### Semi-Auto Weapons Ban
With linear regression `attendance`, `college_graduate` and `non_hisp_black` are also important, while `conservative_protestant` and `catholic` are not. The strongest effect are of sex and race with womens and blacks tending to be in favor of the ban.

#### Ban on High-Capacity Ammo Clips
What is interesting here is the fact that `children` is also an important variable. This implies that underaged children at home influence parents negatively towards this particular ban. Also, `postgraduate` here is not significant. Again sex highly influences the result.

#### Cilivian Hand Gun Possession Ban
`black_protestant` is now also significant and it has a high impact on the result. Being hispanic or of other race also seems to influence the result on a great degree, which was also clear in the logistic regression.

#### Support for Concealed Carry Laws
Statistically significant variables agree with the logistic regression. Again, sex has a large impact on the result. In other words, women are more prone to disagree with concealed carry laws. The same goes for black protestants. The only factors that have a positive contribution are belif in supernatural evil and living in a southern state.

#### More Armed Security at Schools 
A variable that is added in the significant ones is `biblical_literalism`, while `married` is not important. Being postgraduate is a factor that greately affects the result negatively and this may imply that highly educated people are prone to disagree with armed security at schools.

#### More Teachers/Faculty having Guns
In linear regression `biblical_literalism` and `postgraduate` are not important, while `children` is. In fact, people with underaged children are prone to respond that they agree with armed teachers or faculty. Blacks and women are again more prone to disagree.

#### Required Gun Safety Programs
Here, there is not enough evidence that `age` and `hispanic` affect the result, while `less_than_hs` seems to do so: people whose education is less than high school are more prone to disagree with gun safety programs (it has the highest coefficient among important variables). The coefficients generally are low compared to those of other variables.

#### Expanded Mental Health Screening
Statistically significant variables agree with the logistic regression. Again, the coefficients are not high, however we can see that being woman or hispanic are the strongest predictors for agreeing with expanded screening of mental health. 

In [ ]:
from sklearn.model_selection import GridSearchCV, KFold, train_test_split, cross_val_score
from sklearn.tree import DecisionTreeRegressor
from sklearn import tree
import matplotlib.pyplot as plt

In [ ]:
def make_regression_trees(independent):
    for dependent in dependent_variables:
        print("Tree for variable", dependent)


        parameters = [
            {'max_depth': list(range(1, len(independent)))},
            ]
        # Removed n_splits since we have splitted dataset above.
        cv = KFold(shuffle=True, random_state=29)
        clf = GridSearchCV(DecisionTreeRegressor(), parameters, cv=cv)
        clf = clf.fit(x_train, y_train)
        print(clf.best_estimator_)
        print(clf.best_params_)

        gun_tree = DecisionTreeRegressor(max_depth=clf.best_params_["max_depth"])

        gun_tree.fit(x_train, y_train)
        print(gun_tree.score(x_test, y_test))

        plt.figure(figsize=(18, 10))
        tree.plot_tree(gun_tree, 
                    feature_names=selected_vars[independent].columns,
                    filled=True, rounded=True, fontsize=14)
        plt.show(tree)


In [ ]:
make_regression_trees(independent_variables1)

Scores are low. However, we will continue to see scores for all independent variables.

In [ ]:
make_regression_trees(independent_variables2)

Although some scores have improved, we will try random forests to see whether there is room for further improvement.

In [ ]:
def make_random_forests(independent):
    for dependent in dependent_variables:
        print("Tree for variable", dependent)


        parameters = [
            {'max_depth': list(range(1, len(independent)))},
            ]
        # Removed n_splits since we have splitted dataset above.
        cv = KFold(shuffle=True, random_state=29)
        clf = GridSearchCV(DecisionTreeRegressor(), parameters, cv=cv)
        clf = clf.fit(x_train, y_train)
        print(clf.best_estimator_)
        print(clf.best_params_)

        gun_tree = DecisionTreeRegressor(max_depth=clf.best_params_["max_depth"])

        gun_tree.fit(x_train, y_train)
        print(gun_tree.score(x_test, y_test))

        plt.figure(figsize=(18, 10))
        tree.plot_tree(gun_tree, 
                    feature_names=selected_vars[independent].columns,
                    filled=True, rounded=True, fontsize=14)
        plt.show(tree)


In [ ]:
from sklearn.ensemble import RandomForestRegressor

for dependent in dependent_variables:
    print("Forest for variable", dependent)

    forest = RandomForestRegressor(n_estimators=1000, max_depth=None)

    forest.fit(x_train, y_train)

    # Use the forest's predict method on the test data
    predictions = forest.predict(x_test)# Calculate the absolute errors
    errors = abs(predictions - y_test)# Print out the mean absolute error (mae)
    print('Mean Absolute Error:', round(np.mean(errors), 2), 'degrees.')

    scores = cross_val_score(forest, x_test, 
                            y_test, cv=5)
    print(scores.mean())     

In [ ]:
import lightgbm as lgb
from sklearn.metrics import mean_absolute_error



for dependent in dependent_variables:
    print("Forest for variable", dependent)

    lgb_reg = lgb.LGBMRegressor()
    lgb_reg.fit(x_train, y_train, feature_name=independent_variables2)
    preds = lgb_reg.predict(x_test)
    print(mean_absolute_error(y_test, preds))

    scores = cross_val_score(lgb_reg, 
                         selected_vars[independent_variables2], selected_vars[dependent],
                         scoring="neg_mean_absolute_error", cv=5)
    print(-scores)
    print(np.mean(-scores))

    lgb_reg = lgb.LGBMRegressor()
    lgb_reg.fit(selected_vars[independent_variables2], selected_vars[dependent], 
    feature_name=independent_variables2)
    lgb.plot_importance(lgb_reg, figsize=(8, 6))

In [ ]:
import xgboost as xgb
x_train, x_test, y_train, y_test = train_test_split(selected_vars[independent_variables1].values, selected_vars \
    ['teachers_guns'].values, test_size=0.25)
xgb_reg = xgb.XGBRegressor()
xgb_reg.fit(x_train, y_train)
preds = xgb_reg.predict(x_test)
mean_absolute_error(y_test, preds)

In [ ]:
scores = cross_val_score(xgb_reg, 
                         selected_vars[independent_variables1], selected_vars['teachers_guns'], scoring="neg_mean_absolute_error")
print(-scores)
print(np.mean(-scores))

In [ ]:
fig = plt.figure(figsize=(8, 6))
xgb_reg.get_booster().feature_names = independent_variables1
xgb.plot_importance(xgb_reg, ax = fig.gca())